ringtail/storagemanager.py

```diff
@@ -2548,84 +2814,120 @@ def gen(fps):
-                return bs
-
-            bclusters = clusterFps([DataStructs.CreateFromBitString(make_bitstring(pose[2:])) for pose in leff_poseid_ifps], self.interaction_cluster)
-            logger.info(f"Number of interaction fingerprint butina clusters: {len(bclusters)}")
+            # separate out bit strings and do rdkit stuff with it
+            bitstring_list = [
+                DataStructs.CreateFromBitString(item[2]) for item in leff_poseid_ifps
+            ]
+            # calculate clusters
+            bclusters = clusterFps(bitstring_list, self.interaction_cluster)
+            logger.info(
+                f"Number of interaction fingerprint butina clusters: {len(bclusters)}"
+            )
 
             # select ligand from each cluster with best ligand efficiency
             int_rep_poseids = []
             for c in bclusters:
-                c_leffs = np.array([leff_poseid_ifps[i][0] for i in c])  # beware magic numbers
+                # for pose_id in cluster, make array of
+                c_leffs = np.array(
+                    [leff_poseid_ifps[i][0] for i in c]
+                )  # beware magic numbers
                 best_lig_c = leff_poseid_ifps[c[np.argmin(c_leffs)]][1]
                 int_rep_poseids.append(str(best_lig_c))
 
-            self._insert_cluster_data(bclusters, [l[1] for l in leff_poseid_ifps], "ifp", str(self.interaction_cluster))
+            self._insert_cluster_data(
+                bclusters,
+                [l[1] for l in leff_poseid_ifps],
+                "ifp",
+                str(self.interaction_cluster),
+            )  # could this be a problem?
 
             # catch if no pose_ids returned
             if int_rep_poseids == []:
-                logger.warning("No passing results prior to clustering. Clustering not performed.")
+                logger.warning(
+                    "No passing results prior to clustering. Clustering not performed."
+                )
             else:
                 if self.mfpt_cluster is None:
-                    sql_string = output_str + "Pose_ID=" + " OR Pose_ID=".join(int_rep_poseids)
+                    sql_string = (
+                        output_str + "Pose_ID=" + " OR Pose_ID=".join(int_rep_poseids)
+                    )
                 else:
                     unclustered_query = f"SELECT Pose_ID FROM Results WHERE {'Pose_ID=' + ' OR Pose_ID='.join(int_rep_poseids)}"
 
         if self.mfpt_cluster is not None:
-            logger.warning("WARNING: Ligand morgan fingerprint clustering is memory-constrained. Using overly-permissive filters with clustering may cause issues.")# TODO: remove this memory bottleneck
-            logger.warning("N.B.: If using both interaction and morgan fingerprint clustering, the morgan fingerprint clustering will be performed on the results staus post interaction fingerprint clustering.")
+            logger.warning(
```

Comment:
It would be great if we could remove this bottleneck before release. I think I may have gotten this to work in some other code by using generators for the inputs to the RDKit clustering rather than lists?

Reply:
I think that is too big an ask, unless you already know the shape of the solution and it is pretty well defined in which case we can talk about it?